File: commanddelegator/src/main/java/com/greenops/commanddelegator/api/CommandDelegatorApi.java
package com.greenops.commanddelegator.api;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.greenops.util.datamodel.clientmessages.ClientRequest;
import com.greenops.util.dbclient.DbClient;
import com.greenops.util.error.AtlasBadKeyError;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@Slf4j
@RestController
@RequestMapping("/")
public class CommandDelegatorApi {

    private static final String LOCAL_CLUSTER_NAME = "kubernetes_local";

    private final DbClient dbClient;
    private final ObjectMapper objectMapper;

    @Autowired
    public CommandDelegatorApi(DbClient dbClient, ObjectMapper objectMapper) {
        this.dbClient = dbClient;
        this.objectMapper = objectMapper;
    }

    @GetMapping(value = "/requests/{orgName}/{clusterName}")
    public ResponseEntity<List<ClientRequest>> getCommands(@PathVariable("orgName") String orgName, @PathVariable("clusterName") String clusterName) {
        var clusterKey = DbKey.makeDbClusterKey(orgName, clusterName);
        var clusterSchema = dbClient.fetchClusterSchemaTransactionless(clusterKey);
        if (clusterSchema == null) {
            return ResponseEntity.badRequest().build();
        }
        //If no deploy is enabled for the cluster, respond with an empty payload
        if (clusterSchema.getNoDeployInfo() == null) {
            return ResponseEntity.ok().build();
        }
        var notificationQueueKey = DbKey.makeClientNotificationQueueKey(orgName, clusterName);
        var requestQueueKey = DbKey.makeClientRequestQueueKey(orgName, clusterName);
        try {
            //Notifications are from the management plane, they take priority over basic processing
            var clientNotificationHead = dbClient.fetchHeadInClientRequestList(notificationQueueKey);
            if (clientNotificationHead != null) {
                return ResponseEntity.ok().body(List.of(clientNotificationHead));
            }
            var clientRequestPacket = dbClient.fetchHeadInClientRequestList(requestQueueKey);
            if (clientRequestPacket != null) {
                var request = clientRequestPacket.getClientRequest();
                request.setFinalTry(clientRequestPacket.getRetryCount() >= 5);
                return ResponseEntity.ok().body(List.of(request));
            }
        } catch (AtlasBadKeyError err) {
            return ResponseEntity.badRequest().build();
        }
        return ResponseEntity.ok().build();
    }

    @DeleteMapping(value = "/requests/ackHead/{orgName}/{clusterName}")
    public ResponseEntity<Void> ackHeadOfRequestList(@PathVariable("orgName") String orgName,
                                                     @PathVariable("clusterName") String clusterName) {
        var clusterKey = DbKey.makeDbClusterKey(orgName, clusterName);
        if (dbClient.fetchClusterSchemaTransactionless(clusterKey) == null) {
            return ResponseEntity.badRequest().build();
        }
        var key = DbKey.makeClientRequestQueueKey(orgName, clusterName);
        try {
            dbClient.updateHeadInTransactionlessList(key, null);
        } catch (AtlasBadKeyError err) {
            return ResponseEntity.badRequest().build();
        }
        return ResponseEntity.ok().build();
    }

    @PostMapping(value = "/notifications/{orgName}/{clusterName}")
    public ResponseEntity<Void> addNotificationCommand(@PathVariable("orgName") String orgName,
                                                       @PathVariable("clusterName") String clusterName,
                                                       @RequestBody ClientRequest notification) {
        var clusterKey = DbKey.makeDbClusterKey(orgName, clusterName);
        var clusterSchema = dbClient.fetchClusterSchemaTransactionless(clusterKey);
        if (clusterSchema == null) {
            return ResponseEntity.badRequest().build();
        }
        //If no deploy is enabled for the cluster, respond with an empty payload
        if (clusterSchema.getNoDeployInfo() == null) {
            return ResponseEntity.ok().build();
        }
        var key = DbKey.makeClientNotificationQueueKey(orgName, clusterName);
        try {
            dbClient.insertValueInTransactionlessList(key, notification);
        } catch (AtlasBadKeyError err) {
            return ResponseEntity.badRequest().build();
        }
        return ResponseEntity.ok().build();
    }

    @DeleteMapping(value = "/notifications/ackHead/{orgName}/{clusterName}")
    public ResponseEntity<Void> ackHeadOfNotificationList(@PathVariable("orgName") String orgName,
                                                          @PathVariable("clusterName") String clusterName) {
        var clusterKey = DbKey.makeDbClusterKey(orgName, clusterName);
        if (dbClient.fetchClusterSchemaTransactionless(clusterKey) == null) {
            return ResponseEntity.badRequest().build();
        }
        var key = DbKey.makeClientNotificationQueueKey(orgName, clusterName);
        try {
            dbClient.updateHeadInTransactionlessList(key, null);
        } catch (AtlasBadKeyError err) {
            return ResponseEntity.badRequest().build();
        }
        return ResponseEntity.ok().build();
    }

    @DeleteMapping(value = "/requests/retry/{orgName}/{clusterName}")
    public ResponseEntity<Void> retryMessage(@PathVariable("orgName") String orgName,
                                             @PathVariable("clusterName") String clusterName) {
        var clusterKey = DbKey.makeDbClusterKey(orgName, clusterName);
        if (!clusterName.equals(LOCAL_CLUSTER_NAME) && dbClient.fetchClusterSchemaTransactionless(clusterKey) == null) {
            return ResponseEntity.badRequest().build();
        }
        var key = DbKey.makeClientRequestQueueKey(orgName, clusterName);
        try {
            var clientRequestPacket = dbClient.fetchHeadInClientRequestList(key);
            if (clientRequestPacket != null) {
                clientRequestPacket.setRetryCount(clientRequestPacket.getRetryCount() + 1);
            }
            dbClient.insertValueInTransactionlessList(key, clientRequestPacket);
            dbClient.updateHeadInTransactionlessList(key, null);
        } catch (AtlasBadKeyError err) {
            return ResponseEntity.badRequest().build();
        }
        return ResponseEntity.ok().build();
    }
}

File: commanddelegator/command_delegator_api.go
package main

import (
	"bytes"
	"net/http"

	"github.com/greenopsinc/util/kubernetesclient"

	"github.com/google/uuid"
	"github.com/gorilla/mux"
	"github.com/greenopsinc/util/clientrequest"
	"github.com/greenopsinc/util/cluster"
	"github.com/greenopsinc/util/db"
	"github.com/greenopsinc/util/httpserver"
	"github.com/greenopsinc/util/serializer"
	"github.com/greenopsinc/util/serializerutil"
	"github.com/greenopsinc/util/starter"
	"github.com/greenopsinc/util/tlsmanager"
)

const (
	orgNameField     string = "orgName"
	clusterNameField string = "clusterName"
)

var dbOperator db.DbOperator
var emptyClusterStruct = cluster.ClusterSchema{}
var emptyClientRequestPacketStruct = clientrequest.ClientRequestPacket{}

func getCommands(w http.ResponseWriter, r *http.Request) {
	vars := mux.Vars(r)
	orgName := vars[orgNameField]
	clusterName := vars[clusterNameField]
	dbClient := dbOperator.GetClient()
	defer dbClient.Close()

	clusterKey := db.MakeDbClusterKey(orgName, clusterName)
	clusterSchema := dbClient.FetchClusterSchemaTransactionless(clusterKey)
	if clusterSchema == emptyClusterStruct {
		http.Error(w, "Cluster was nil", http.StatusBadRequest)
		return
	}
	noDeployInfo := clusterSchema.NoDeploy
	notificationQueueKey := db.MakeClientNotificationQueueKey(orgName, clusterName)
	requestQueueKey := db.MakeClientRequestQueueKey(orgName, clusterName)

	//Notifications are from the management plane, they take priority over basic processing
	//Notifications also do not change the state of deployments and clusters, so can be processed even when in a no-deploy state
	clientNotificationHead := dbClient.FetchHeadInClientRequestList(notificationQueueKey)
	if clientNotificationHead != emptyClientRequestPacketStruct {
		//Notifications don't subscribe to the retry/final try logic, the client wrapper makes sure it's only run once
		writeResponse(w, []clientrequest.ClientRequestEvent{clientNotificationHead.ClientRequest})
		return
	}
	if noDeployInfo != nil && noDeployInfo.Namespace == "" {
		writeResponse(w, nil)
		return
	}
	clientRequestPacket := dbClient.FetchHeadInClientRequestList(requestQueueKey)
	if clientRequestPacket != emptyClientRequestPacketStruct {
		if noDeployInfo != nil && clientRequestPacket.Namespace == noDeployInfo.Namespace {
			writeResponse(w, nil)
			return
		}
		request := clientRequestPacket.ClientRequest
		request.SetFinalTry(clientRequestPacket.RetryCount >= 5)
		writeResponse(w, []clientrequest.ClientRequestEvent{request})
		return
	}
}

func ackHeadOfRequestList(w http.ResponseWriter, r *http.Request) {
	vars := mux.Vars(r)
	orgName := vars[orgNameField]
	clusterName := vars[clusterNameField]
	dbClient := dbOperator.GetClient()
	defer dbClient.Close()

	clusterKey := db.MakeDbClusterKey(orgName, clusterName)
	clusterSchema := dbClient.FetchClusterSchemaTransactionless(clusterKey)
	if clusterSchema == emptyClusterStruct {
		http.Error(w, "Cluster was nil", http.StatusBadRequest)
		return
	}
	key := db.MakeClientRequestQueueKey(orgName, clusterName)
	dbClient.UpdateHeadInTransactionlessList(key, nil)
}

func addNotificationCommand(w http.ResponseWriter, r *http.Request) {
	vars := mux.Vars(r)
	orgName := vars[orgNameField]
	clusterName := vars[clusterNameField]
	dbClient := dbOperator.GetClient()
	defer dbClient.Close()

	clusterKey := db.MakeDbClusterKey(orgName, clusterName)
	clusterSchema := dbClient.FetchClusterSchemaTransactionless(clusterKey)
	if clusterSchema == emptyClusterStruct {
		http.Error(w, "Cluster was nil", http.StatusBadRequest)
		return
	}
	key := db.MakeClientNotificationQueueKey(orgName, clusterName)
	requestId := uuid.New().String()

	buf := new(bytes.Buffer)
	_, err := buf.ReadFrom(r.Body)
	if err != nil {
		http.Error(w, "Error when reading request body", http.StatusBadRequest)
		return
	}
	notification := serializer.Deserialize(buf.String(), serializerutil.ClientRequestType).(clientrequest.ClientRequestEvent)
	notification.(clientrequest.NotificationRequestEvent).SetRequestId(requestId)
	dbClient.InsertValueInTransactionlessList(key, clientrequest.ClientRequestPacket{
		Namespace:     "",
		ClientRequest: notification,
	})

	writeResponse(w, requestId)
}

func ackHeadOfNotificationList(w http.ResponseWriter, r *http.Request) {
	vars := mux.Vars(r)
	orgName := vars[orgNameField]
	clusterName := vars[clusterNameField]
	dbClient := dbOperator.GetClient()
	defer dbClient.Close()

	clusterKey := db.MakeDbClusterKey(orgName, clusterName)
	clusterSchema := dbClient.FetchClusterSchemaTransactionless(clusterKey)
	if clusterSchema == emptyClusterStruct {
		http.Error(w, "Cluster was nil", http.StatusBadRequest)
		return
	}
	key := db.MakeClientNotificationQueueKey(orgName, clusterName)
	dbClient.UpdateHeadInTransactionlessList(key, nil)
}

func retryMessage(w http.ResponseWriter, r *http.Request) {
	vars := mux.Vars(r)
	orgName := vars[orgNameField]
	clusterName := vars[clusterNameField]
	dbClient := dbOperator.GetClient()
	defer dbClient.Close()

	clusterKey := db.MakeDbClusterKey(orgName, clusterName)
	clusterSchema := dbClient.FetchClusterSchemaTransactionless(clusterKey)
	if clusterSchema == emptyClusterStruct {
		http.Error(w, "Cluster was nil", http.StatusBadRequest)
		return
	}
	key := db.MakeClientRequestQueueKey(orgName, clusterName)
	clientRequestPacket := dbClient.FetchHeadInClientRequestList(key)
	if clientRequestPacket != emptyClientRequestPacketStruct {
		clientRequestPacket.RetryCount = clientRequestPacket.RetryCount + 1
	}
	dbClient.InsertValueInTransactionlessList(key, clientRequestPacket)
	dbClient.UpdateHeadInTransactionlessList(key, nil)
}

func writeResponse(w http.ResponseWriter, i interface{}) {
	if i != nil {
		payload := serializer.Serialize(i)
		w.Write([]byte(payload))
	}
	w.Header().Set("Content-Type", "application/json")
}

func InitEndpoints(r *mux.Router) {
	r.HandleFunc("/requests/{orgName}/{clusterName}", getCommands).Methods("GET")
	r.HandleFunc("/requests/ackHead/{orgName}/{clusterName}", ackHeadOfRequestList).Methods("DELETE")
	r.HandleFunc("/notifications/{orgName}/{clusterName}", addNotificationCommand).Methods("POST")
	r.HandleFunc("/notifications/ackHead/{orgName}/{clusterName}", ackHeadOfNotificationList).Methods("DELETE")
	r.HandleFunc("/requests/retry/{orgName}/{clusterName}", retryMessage).Methods("DELETE")
}

func main() {
	dbOperator = db.New(starter.GetDbClientConfig())
	var tlsManager tlsmanager.Manager
	var kubernetesClient kubernetesclient.KubernetesClient
	kubernetesClient = kubernetesclient.New()

	if starter.GetNoAuthClientConfig() {
		tlsManager = tlsmanager.NoAuth()
	} else {
		tlsManager = tlsmanager.New(kubernetesClient)
	}
	r := mux.NewRouter()
	InitEndpoints(r)
	httpserver.CreateAndWatchServer(tlsmanager.ClientCommandDelegator, tlsManager, r)
}
